src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -737,28 +850,27 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
             filter->get_policy_driver()->get_cache_policy()->update(dpp, oid, ofs, bl.length(), version, *y);
 
 	    /* Store block in directory */
-            if (!blockDir->exist_key(&block, *y)) {
-              if (blockDir->set(&block, *y) < 0) //should we revert previous steps if this step fails?
-		ldpp_dout(dpp, 10) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory set method failed." << dendl;
+	    if (blockDir->set(dpp, &block, true, *y) < 0) { //should we revert previous steps if this step fails?
```

Comment:
Here, what essentially is being done is that we set a block with the new values. This piece of code, first checks if the key doesn't exist then we set the values as is. If the key does exist, we get the old values and check the version., if there is a mismatch then we delete the old block and set the new one, and if there is no version mismatch we just update the host.

The above can be modified as:
1. Get the block (note: that exist_key should be removed from get) -> if this returns -ENOENT, it means the block does not exist
2. I 1. returns -ENOENT, then set the values as is.
3. If 1. returns success, then check for version field, if there is a mismatch, then update the version field
4. Also update the hostsList withe the address of the local cache.
5. Call set with all updated values. (Now here you can see that a get has already been invoked, so all the old values will be retained, and you can do without the exist_key check in set)